File: app/src/main/java/com/github/willjgriff/ethereumwallet/ui/transactions/mvp/TransactionsPresenter.java
package com.github.willjgriff.ethereumwallet.ui.transactions.mvp;

import com.github.wiljgriff.ethereumwallet.ethereum.node.Ethereum;
import com.github.willjgriff.ethereumwallet.mvp.BaseMvpPresenter;

import javax.inject.Inject;

import timber.log.Timber;

/**
 * Created by Will on 29/01/2017.
 */

public class TransactionsPresenter extends BaseMvpPresenter<TransactionsView> {

	private Ethereum mEthereum;

	@Inject
	public TransactionsPresenter(Ethereum ethereum) {
		mEthereum = ethereum;
	}

	@Override
	public void viewReady() {
		addDisposable(mEthereum.getBlockHeaderObservable()
			.subscribe(header -> getView().newHeader(header),
				throwable -> Timber.e(throwable)));

		addDisposable(mEthereum.getPeersInfo()
			.subscribe(peerInfos -> getView().updatePeerInfos(peerInfos),
				throwable -> Timber.e(throwable)));
	}
}

File: app/src/main/kotlin/com/github/willjgriff/ethereumwallet/ui/transactions/mvp/TransactionsPresenter.kt
package com.github.willjgriff.ethereumwallet.ui.transactions.mvp

import com.github.willjgriff.ethereumwallet.ethereum.transactions.TransactionsManager
import com.github.willjgriff.ethereumwallet.mvp.BaseMvpPresenter
import io.reactivex.Observable
import io.reactivex.disposables.Disposable
import javax.inject.Inject

/**
 * Created by williamgriffiths on 11/04/2017.
 */
class TransactionsPresenter @Inject constructor(private val transactionsManager: TransactionsManager) : BaseMvpPresenter<TransactionsView>() {

    private lateinit var transactionsDisposable: Disposable

    fun setObservables(clearTransactions: Observable<Any>, selectSearchRange: Observable<Any>) {
        setupClearTransactions(clearTransactions)
    }

    override fun viewReady() {
        subscribeToTransactions()
    }

    private fun subscribeToTransactions() {
        transactionsDisposable = transactionsManager
                .transactionsObservable.subscribe { view.addTransaction(it) }
        addDisposable(transactionsDisposable)
    }

    private fun setupClearTransactions(clearTransactions: Observable<Any>) {
        clearTransactions.subscribe {
            view.clearTransactions()
            transactionsManager.clearAndRestart()
            transactionsDisposable.dispose()
            subscribeToTransactions()
        }
    }
}